Skins: allow custom brightness threshold for hotcue colors

ronso0 opened the pull request:
The style of WHotcueButton can be adjusted for dim & bright hotcue colors.
In LateNight PaleMoon theme which is rather dim the default brightness threshold `127` doesn't work well; colors are considered 'dark' too early.

This change allows to set a custom `<DimBrightThreshold>` (int 0-256) for the following widgets:
* HotcueButton
* Overview
* Visual (scrolling waveforms)

In the PaleMoon example I used the default 127 in the overview to keep the contrasting marker strokes.
(see also #3003)

ronso0 committed:
color util: rename isDimmColor > is DimColor, add 'b' to bool variables

ronso0 commented:
Now waveforms and overviews can read the `<DimBrightThreshold>` node, too. Default is 127.

The sole purpose is to adjust cue colors to dim PaleMoon theme. Please test!

Holzhaus commented on src/waveform/renderers/waveformsignalcolors.h:
```diff
@@ -25,6 +25,9 @@ class WaveformSignalColors {
     inline const QColor& getPlayedOverlayColor() const { return m_playedOverlayColor; }
     inline const QColor& getPassthroughOverlayColor() const { return m_passthroughOverlayColor; }
     inline const QColor& getBgColor() const { return m_bgColor; }
+    inline const int& getDimBrightThreshold() const {
```
Don't use a const reference, just return int.

ronso0 replied:
just `inline int getDimBrightThreshold() const {` then?

Holzhaus replied:
Yes.

Holzhaus commented on src/widget/wwaveformviewer.cpp:
```diff
@@ -27,7 +27,8 @@ WWaveformViewer::WWaveformViewer(
           m_pCueMenuPopup(make_parented<WCueMenuPopup>(pConfig, this)),
-          m_waveformWidget(nullptr) {
+          m_waveformWidget(nullptr),
+          m_dimBrighTreshold(127) {
```
```suggestion
          m_dimBrightTreshold(127) {
```
and everywhere else

ronso0 replied:
nice catch.

Holzhaus reviewed (changes requested):
Thanks. I can't test properly, because it hangs very long when starting and loading a track doesn't work either, but I doubt this was caused by your changes. Probably this is because this branch is so old. I suggest you rebase on latest 2.3 and also change the target branch to 2.3.

ronso0 commented:
> Thanks. I can't test properly, because it hangs very long when starting and loading a track doesn't work either, but I doubt this was caused by your changes. Probably this is because this branch is so old. I suggest you rebase on latest 2.3 and also change the target branch to 2.3.

This is based on 2.3 already and FWIW I didn't encounter the issues you did.
I rebased and targeted at 2.3

Holzhaus commented:
> > Thanks. I can't test properly, because it hangs very long when starting and loading a track doesn't work either, but I doubt this was caused by your changes. Probably this is because this branch is so old. I suggest you rebase on latest 2.3 and also change the target branch to 2.3.
> 
> This is based on 2.3 already and FWIW I didn't encounter the issues you did.
> I rebased and targeted at 2.3

Yes, I got some weird ACPI issues on my new laptop. A reboot fixed them.

This is how it looks on my machine. I think we need light hotcue button labels, but other than it it looks fine.

![Screenshot from 2020-08-17 23-01-13](https://user-images.githubusercontent.com/1834516/90444511-413b7a80-e0de-11ea-8625-a2e4093931cc.png)

ronso0 commented:
...and again the SVGs are rendered diferently it seems:
![image](https://user-images.githubusercontent.com/5934199/90447927-ac884b00-e0e4-11ea-8857-421e43e03aea.png)

now that it works I can increase the blur, but I'm a bit tired of this cross-computer testing by now, so I may as well switch to bright icons, even if that's what I wanted to avoid in the first place.

Thanks for testing!

Holzhaus commented:
> ...and again the SVGs are rendered diferently it seems:
> ![image](https://user-images.githubusercontent.com/5934199/90447927-ac884b00-e0e4-11ea-8857-421e43e03aea.png)
> 
> now that it works I can increase the blur, but I'm a bit tired of this cross-computer testing by now, so I may as well switch to bright icons, even if that's what I wanted to avoid in the first place.

Whoops, the low contrast might actually not be Qt's fault. It's already late and I forgot that I have redshift enabled on my computer, so this might have caused this. Anyway, I think the contrast is a bit too low especially when you turn down screen brightness in a club setting, so using white labels or conderably more blur would probably be a good idea. I'd probably prefer white text instead of a strong blue effect.

ronso0 commented:
Okay, good to know :)
I'll play around with a bit, maybe bright icons are more cnsistent in the end considering the waveform labels also switch to bright fonts. I'm just glad it doesn't pick a bright text for the orange in that picture, that was really a hard contrast at night.

daschuer commented on src/waveform/renderers/waveformsignalcolors.cpp:
```diff
@@ -76,6 +76,7 @@ bool WaveformSignalColors::setup(const QDomNode &node, const SkinContext& contex
     }
     m_bgColor = WSkinColor::getCorrectColor(m_bgColor).toRgb();
 
+    m_dimBrightThreshold = context.selectInt(node, QStringLiteral("DimBrightThreshold"));
```
I think we should do here something like this to have the value optional: 
```  
    bool okay = false;
    m_dimBrightThreshold = context.selectInt(node, QStringLiteral("DimBrightThreshold"), &okay);
    if (okay) {
        m_dimBrightThreshold = kDefaulDimBrightThreshold;
    }
```

daschuer commented on src/waveform/renderers/waveformsignalcolors.h:
```diff
@@ -45,6 +48,7 @@ class WaveformSignalColors {
     QColor m_playedOverlayColor;
     QColor m_passthroughOverlayColor;
     QColor m_bgColor;
+    int m_dimBrightThreshold = 127;
```
We avoid initialize value in the class definition, because this introduces a hidden static constructor.  
Please move this to the constructor and use kDefaulDimBrightThreshold.

daschuer commented on src/waveform/renderers/waveformwidgetrenderer.cpp:
```diff
@@ -262,6 +263,8 @@ void WaveformWidgetRenderer::setup(
         m_orientation = Qt::Horizontal;
     }
 
+    m_dimBrightTreshold = context.selectInt(node, QStringLiteral("DimBrightThreshold"));
```
The same here. Please default to 127

daschuer commented on src/widget/whotcuebutton.cpp:
```diff
@@ -12,9 +12,10 @@ WHotcueButton::WHotcueButton(const QString& group, QWidget* pParent)
-          m_cueColorDimmed(false),
-          m_isCueColorLight(false),
-          m_isCueColorDark(false) {
+          m_cueColorDimThreshold(127),
```
This should become a const in the anonymous namespace

daschuer commented on src/widget/whotcuebutton.cpp:
```diff
@@ -28,6 +29,7 @@ void WHotcueButton::setup(const QDomNode& node, const SkinContext& context) {
     } else {
         SKIN_WARNING(node, context) << "Hotcue value invalid";
     }
+    m_cueColorDimThreshold = context.selectInt(node, QStringLiteral("DimBrightThreshold"));
```
This should not keep the default if the value is not defined in skins.

ronso0 replied:
done.

daschuer commented on src/widget/whotcuebutton.cpp:
```diff
@@ -132,13 +134,13 @@ void WHotcueButton::slotColorChanged(double color) {
         // Since we don't know the color reset both
-        m_isCueColorLight = false;
-        m_isCueColorDark = false;
+        m_bCueColorIsLight = false;
```
In Mixxx we have no prefix for bools. The old version was IMHO fine.

daschuer reviewed (changes requested):
Thank you for this useful PR. I have left some comments.

ronso0 commented:
Just to let you know: I addressed the comments but didn't push, yet, because I need some continous time to thoroughly test the thresholds with updated hotcue icons and the waveforms.
This won't happen before next week, though.

ronso0 committed:
WHotcueButton/color.h: allow custom dim/bright threshold

ronso0 committed:
WOverView/WVisual: allow custom dim/bright treshold for marks

ronso0 committed:
LateNight PaleMoon: set custom dim/bright thresholds for waveform marks and hotcues, refine light hotcue icons

ronso0 commented:
A TMP commit slipped in so I finally cleaned up and rebased onto 2.3. Please have a look!

I adjusted the thresholds in LateNight PaleMoon and made the hotcue button icons more clear.
The comparison uses the new icons for demonstration.
before (general threshold = 127):
![Screenshot_2020-09-23_15-27-19](https://user-images.githubusercontent.com/5934199/94022420-5da19580-fdb5-11ea-9b73-cb229132c657.png)

this PR (custom threshold for buttons and waveform = 90):
![Screenshot_2020-09-23_15-27-19___90](https://user-images.githubusercontent.com/5934199/94022436-62664980-fdb5-11ea-97ac-22e1f43a0c56.png)

Holzhaus commented:
90 is okay, 80 makes the "6" barely readable.

Holzhaus reviewed (commented):
There are merge conflicts.

Also, I think the contrast for the blue hotcue colors from the default palette is bad. And maybe the red and purple button should be considered "dark", too?

![Screenshot from 2020-10-18 16-31-07](https://user-images.githubusercontent.com/1834516/96370867-9daf2a80-115f-11eb-934e-807a04ff6140.png)

ronso0 commented:
you can try a different threshold, set in skin.xml in PaleMoon color scheme

Holzhaus commented:
> There are merge conflicts.
> 
> Also, I think the contrast for the blue hotcue colors from the default palette is bad. And maybe the red and purple button should be considered "dark", too?
> 
> ![Screenshot from 2020-10-18 16-31-07](https://user-images.githubusercontent.com/1834516/96370867-9daf2a80-115f-11eb-934e-807a04ff6140.png)

I tried, using the light grey text on red make the contrast even worse.

ronso0 committed:
Merge branch '2.3' into hotcue-dim-bright

ronso0 commented:
I'll check if it can be fixed with a brighter font for dark buttons.

ronso0 commented:
Oh, the first time for quite a while I got a "Build passed" message from our friend Travis :)

ronso0 commented:
> I think the contrast for the blue hotcue colors from the default palette is bad

I made the 'dark' icons a bit brighter:
![image](https://user-images.githubusercontent.com/5934199/96655204-3949c200-133d-11eb-8fee-1849b442f71e.png)

> And maybe the red and purple button should be considered "dark", too?

I don't want to tweak our dark/light algorythm here. Would you mind fiing a bug report?

ronso0 committed:
Latenight PaleMoon: brighter icons for dark hotcues

Holzhaus commented:
> > And maybe the red and purple button should be considered "dark", too?
> 
> I don't want to tweak our dark/light algorythm here. Would you mind fiing a bug report?

The algorithm is okay, I tweaked the threshold to `105` and now it looks like this:

![Screenshot from 2020-10-21 11-55-00](https://user-images.githubusercontent.com/1834516/96704662-a2a8ef80-1394-11eb-8e64-716ee3455cf5.png)

I think that has better contrast, but you decide. Just let me know, so we can merge this.

daschuer reviewed (approved):
Thank you.

ronso0 commented:
> > > And maybe the red and purple button should be considered "dark", too?
> > 
> > 
> > I don't want to tweak our dark/light algorythm here. Would you mind fiing a bug report?
> 
> The algorithm is okay, I tweaked the threshold to `105` and now it looks like this:

I get the impression what I'm trying to improve here is not matter of the dim/bright threshold, but it also seems to depend a lot on the screen used and _maybe also_ the algorhythm, because when I tested with Mixxx Hotcue palette yesterday I perceived the Red indeed darker than the Blue.

I did check again with https://webaim.org/resources/contrastchecker/ and you're right: with the new threshold we get ~4.1-3 for the Red, Blue, Purple which is considered okay for normal text.

Did you try `105` also for the waveforms?
If you find that value working for both Hotcues and waveforms, let me know.

Holzhaus commented:
> Did you try `105` also for the waveforms?
> If you find that value working for both Hotcues and waveforms, let me know.

Yes, looks good to me.

ronso0 committed:
LateNight: adjust dim/bright threshold to 105

ronso0 commented:
Thanks!
Are we done here?

Holzhaus reviewed (changes requested):
There are some pre-commit issues, please check.

ronso0 commented:
those are about
```diff
-    int getDimBrightThreshold() { return m_dimBrightThreshold; }
+    int getDimBrightThreshold() {
+        return m_dimBrightThreshold;
+    }
```
Either I reformat all surrounding one-liners or we ignore the complaint. I don't care but I'd like to have some consitency in that file.
(IIRC we already discussed that specific spot and agreed to ignore, but I'm not sure)

Holzhaus commented:
I'm okay with reformatting all of them.

ronso0 committed:
waveformsignalcolors.h/waveformwidgetrenderer.h: break one-line get functions

ronso0 commented:
Ready.

Holzhaus commented on src/waveform/renderers/waveformwidgetrenderer.h:
```diff
@@ -74,27 +82,54 @@ class WaveformWidgetRenderer {
+    double getZoomFactor() const {
+        return m_zoomFactor;
+    }
+    double getGain() const {
+        return m_gain;
+    }
```
Not your fault, but shouldn't this be `CSAMPLE_GAIN` instead of `double`? Anyway, fixing this is out of scope for this PR.

Holzhaus reviewed (approved):
LGTM, thank you
